Large blocks of replica client output buffer could lead to psync loops and unnecessary memory usage

xbasel opened the pull request:
This can happen when a key almost equal or larger than the
client output buffer limit of the replica is written.

Example:
1. DB is empty
2. Backlog size is 1 MB
3. Client out put buffer limit is 2 MB
4. Client writes a 3 MB key
5. The shared replication buffer will have a single node which contains
the key written above, and it exceeds the backlog size.

At this point the client output buffer usage calculation will report the
replica buffer to be 3 MB (or more) even after sending all the data to
the replica.
The primary drops the replica connection for exceeding the limits,
the replica reconnects and successfully executes partial sync but the
primary will drop the connection again because the buffer usage is still
3 MB. This happens over and over.

To mitigate the problem, this fix limits the maximum size of a single
backlog node to be (repl_backlog_size/16). This way a single node can't
exceed the limits of the COB (the COB has to be larger than the
backlog).
It also means that if the backlog has some excessive data it can't trim,
it would be at most about 6% overuse.

other notes:
1. a loop was added in feedReplicationBuffer which caused a massive LOC
  change due to indentation, the actual changes are just the `min(max` and the loop.
3. an unrelated change in an existing test to speed up a server termination which took 10 seconds.

sundb commented:
LGTM, Can you make some tests to cover it?

oranagra reviewed (approved):
LGTM. but i'll ask @ShooterIT to have a look as well.

hpatro reviewed (commented):
There are bunch of replication related tests under `tests/integration/`, Shall we put this test along with those ?

xbasel commented:
> There are bunch of replication related tests under `tests/integration/`, Shall we put this test along with those ?

Not sure if this test qualifies as an integration test.

ShooterIT commented:
before 7.0, we also think the allocated memory is ClientOutputBufferMemoryUsage, but backlog doesn't share replication buffer with replica clients, so there is no this problem, right?

current fix is ok, but i think we also should care about the calls of this function, such as `getClientMemoryUsage`, maybe it is not accurate after applying this patch.

xbasel commented:
> before 7.0, we also think the allocated memory is ClientOutputBufferMemoryUsage, but backlog doesn't share replication buffer with replica clients, so there is no this problem, right?
Unless client-output-buffer limits are smaller than backlog, there's no problem because the backlog and  replicas do not share the same buffer. Worst case when a big key is written would trigger full sync.

> current fix is ok, but i think we also should care about the calls of this function, such as `getClientMemoryUsage`, maybe it is not accurate after applying this patch.


`getClientMemoryUsage` won't be accurate. But the inaccuracy will be only in the calculation of the head and tail of the relevant  shared replication buffer nodes.  This might affect client eviction (if the keys stored in the head are proportionally large compared to the memory). 

We can have separate calculation for client eviction, but this is an over. IMO this fix is good enough.

oranagra commented on tests/unit/replica-cob-usage.tcl:
```diff
@@ -0,0 +1,32 @@
+
+    test "replica client output buffer usage: client output buffer usage shouldn't include the size of the data it has already consumed" {
+        $slave slaveof $master_host $master_port
+	set keysize [expr 1024*1024]
```
```suggestion
        set keysize [expr 1024*1024]
```

oranagra commented on tests/test_helper.tcl:
```diff
@@ -94,6 +94,7 @@ set ::all_tests {
     unit/client-eviction
     unit/violations
     unit/replybufsize
+    unit/replica-cob-usage
```
i think we can move this to either `tests/unit/obuf-limits.tcl` or `tests/integration/replication-buffer.tcl`, doesn't deserve a file of its own.

oranagra commented on src/networking.c:
```diff
@@ -3625,7 +3625,7 @@ size_t getClientOutputBufferMemoryUsage(client *c) {
             replBufBlock *last = listNodeValue(listLast(server.repl_buffer_blocks));
             replBufBlock *cur = listNodeValue(c->ref_repl_buf_node);
-            repl_buf_size = last->repl_offset + last->size - cur->repl_offset;
+            repl_buf_size = last->repl_offset + last->used - (cur->repl_offset+c->ref_block_pos);
```
it's a little odd that for normal clients we count the memory usage (i.e. the memory their buffers consume, even the unused portions), but for replica we count only the used portion.

e.g. for a normal client we'll deduct `reply_bytes` only when releasing a reply buffer node, not as soon as part of it was written to the socket.

on the other hand replicas share memory with each other, and with the backlog, so a node can be retained even if the replica doesn't need it anymore at all. although i think that's true only for the `last` node, which can still be appended to.

maybe the root of the problem here is that we allow very large nodes, and it's wrong to do that for replications buffers that are both shared, and also retained even when drained (i.e. for the backlog).
maybe we just need to limit their maximum size?

oranagra replied:
in fact, there's another place where if the backlog size is small, and the replication buffers are big, we attribute all the bytes beyond the configured backlog size to the replicas, but in fact, it could just be a huge node that's retained.
so that area can be problematic for very large replication buffers nodes, and will get solved as well if we limit their size.

https://github.com/redis/redis/blob/eef29b68a2cd94de1f03aa1b7891af75f5cabae2/src/evict.c#L354-L364

oranagra commented on tests/unit/replica-cob-usage.tcl:
```diff
@@ -0,0 +1,32 @@
+    set master_host [srv -1 host]
+    set master_port [srv -1 port]
+
+    test "replica client output buffer usage: client output buffer usage shouldn't include the size of the data it has already consumed" {
```
since the main concern we have is about obuf limit disconnection, maybe we should test that instead of just the info metric?
i.e. it could happen that one day the code affecting the obuf limit disconnection is change to break this concern, but he info metric remains unaffected and this test will still pass.

oranagra reviewed (commented):
> Unless client-output-buffer limits are smaller than the backlog, there's no problem because the backlog and replicas do not share the same buffer. Worst case when a big key is written would trigger full sync.

this configuration doesn't make sense, so we have this in place (i.e. the effective replica buffer limit is the bigger of the two):
https://github.com/redis/redis/blob/eef29b68a2cd94de1f03aa1b7891af75f5cabae2/src/networking.c#L3705-L3714

oranagra commented:
i'm sorry, only now i had enough time to dive into this issue and think about it.
i'm not comfortable with the solution, both because of the fact that it counts only used bytes, and not memory consumption, and also because in that case of huge nodes, we have other issues.
i'm not certain what's the right move, hoping to hear some ideas and views.

xbasel commented:
> i'm sorry, only now i had enough time to dive into this issue and think about it. i'm not comfortable with the solution, both because of the fact that it counts only used bytes, and not memory consumption, and also because in that case of huge nodes, we have other issues. i'm not certain what's the right move, hoping to hear some ideas and views.

If memory consumption for a single node is a concern, we probably shouldn't allocate big nodes. Maybe we should limit node size to `PROTO_REPLY_CHUNK_BYTES`.

oranagra commented:
even when the backlog was a plain pre-allocated cyclic buffer, it could be that it's beginning represents the end of the previous command and it was useless, i don't think we wanna worry about that.

but instead, with the new mechanism. we could be aiming to a certain size of backlog, and miss it by a lot because the nodes are too big, affecting our trimming granularity.
and also as you originally pointed out, it can cause replicas to get disconnected due to obuf limit.

i could try to argue that any excess memory we can attribute to the backlog rather than the replicas, and use that excuse to avoid disconnecting them, but the fact is that the backlog config must be set to lower than the replicas obuf limit (i posted the code that's doing that in a previous comment).
and also as i noted i think the metrics should show consumed memory (with any implied overheads), and not just the used net.

do you guys think that just limiting the node side to a reasonable scale (maybe backlog size / 10 or such) can be a reasonable solution to all these issues?
@ShooterIT @sundb please share your thoughts too.

xbasel commented:
> even when the backlog was a plain pre-allocated cyclic buffer, it could be that it's beginning represents the end of the previous command and it was useless, i don't think we wanna worry about that.
> 
> but instead, with the new mechanism. we could be aiming to a certain size of backlog, and miss it by a lot because the nodes are too big, affecting our trimming granularity. and also as you originally pointed out, it can cause replicas to get disconnected due to obuf limit.
> 
> i could try to argue that any excess memory we can attribute to the backlog rather than the replicas, and use that excuse to avoid disconnecting them, but the fact is that the backlog config must be set to lower than the replicas obuf limit (i posted the code that's doing that in a previous comment). and also as i noted i think the metrics should show consumed memory (with any implied overheads), and not just the used net.
> 
> do you guys think that just limiting the node side to a reasonable scale (maybe backlog size / 10 or such) can be a reasonable solution to all these issues? @ShooterIT @sundb please share your thoughts too.

Why to limit the node size to backlog/10 and not to keep it simple and limit the node size to the default  node size `PROTO_REPLY_CHUNK_BYTES` which is 16KB ? the overhead isn't going to be significant.

sundb commented:
> Why to limit the node size to backlog/10 and not to keep it simple and limit the node size to the default node size `PROTO_REPLY_CHUNK_BYTES` which is 16KB ? the overhead isn't going to be significant.

The min value of repl_backlog_size is 16k which will be problematic at this time.
`backlog_size/10` would seem more sense.

xbasel commented:
> > Why to limit the node size to backlog/10 and not to keep it simple and limit the node size to the default node size `PROTO_REPLY_CHUNK_BYTES` which is 16KB ? the overhead isn't going to be significant.
> 
> The min value of repl_backlog_size is 16k which will be problematic at this time. `backlog_size/10` would seem more sense.

why `PROTO_REPLY_CHUNK_BYTES` is problematic ?

sundb commented:
> why `PROTO_REPLY_CHUNK_BYTES` is problematic ?

If we use `PROTO_REPLY_CHUNK_BYTES`, the allocation size of replBufBlock will be 20k(16K+sizeof(replBufBlock)), slave's memory usage may still exceed backlog size.

oranagra commented:
@sundb you're referring to a case where the backlog size is set to lower than 20k?
note that either way we'll use the internal frag, so aiming for 16k we'll still consume (and use) 20.

but anyway, i don't think we need that level of granularity.
considering the replica obuf limit is bigger than the backlog, we can focus on the size of the backlog.
so if nodes are limited to 1/10 of the backlog size, at most the backlog will keep 110% of what it was meant to hold.
[on a second thought, 10 is such an ugly number... better use 8 or 16]

so the question remains, if that'll solve all problems, or do we have concerns that this will not solve?

xbasel commented:
> @sundb you're referring to a case where the backlog size is set to lower than 20k? note that either way we'll use the internal frag, so aiming for 16k we'll still consume (and use) 20.
> 
> but anyway, i don't think we need that level of granularity. considering the replica obuf limit is bigger than the backlog, we can focus on the size of the backlog. so if nodes are limited to 1/10 of the backlog size, at most the backlog will keep 110% of what it was meant to hold. [on a second thought, 10 is such an ugly number... better use 8 or 16]
> 
> so the question remains, if that'll solve all problems, or do we have concerns that this will not solve?

I've limited node size to backlog_size/16

oranagra commented:
@xbasel it took me some time to re-read the discussion and remember what this PR is about..
i see you implemented my suggested solution, i'll review the new code shortly (sorry for the delay, i was busy elsewhere)

i went over the unresolved comments and marked them all as resolved since they're no longer relevant.
i see the top comment is outdated, and i notice you force-pushed a commit with an updated comment (i'll copy it to the top comment).
p.s. i see that force-push contains both a rebase, and a completely new approach (compared to the previous one).
i generally ask to avoid force-push (makes it harder to do incremental review), but since this is a completely new approach, and new tests (i think), that's ok.

oranagra commented:
@xbasel due to the indentation change of a mass of code, it was quite impossible to review the PR in GH, and i ended up cloning it for a local review.
after i adjusted the indentation for my review, i noticed a few unintentional changes probably made by your editor, so i reverted them.
i also fixed a bug (i think) which would have caused us to use `repl_backlog_size / 16` even if we only require smaller size.

i was trying to find a solution to that indentation problem, and avoid massive blame-log destruction (difficulty for future reviewers, and possibly more merge conflicts than necessary), but other than using the notorious `goto`, i can't find one, so i accepted the change.
i'll note that in the top comment.

other than that i did some other cleanups:
1. added comment in the code explaining it's logic.
2. remove unnecessary comment about internal implementation in the test

since i already cloned the code and run the test locally, i did some other cleanup changes in the tests:
* avoiding the use of "slave" when possible in new code
* generalize client_field proc
* move that test out of the scope of the server of the previous test.
* speed up a previous test which i notices takes forever to finish.
* negated the comment saying "and server cron to run and disconnect" (correct me if i'm wrong)

let me know if you see anything wrong.

xbasel commented:
> @xbasel due to the indentation change of a mass of code, it was quite impossible to review the PR in GH, and i ended up cloning it for a local review. after i adjusted the indentation for my review, i noticed a few unintentional changes probably made by your editor, so i reverted them. i also fixed a bug (i think) which would have caused us to use `repl_backlog_size / 16` even if we only require smaller size.
> 
> i was trying to find a solution to that indentation problem, and avoid massive blame-log destruction (difficulty for future reviewers, and possibly more merge conflicts than necessary), but other than using the notorious `goto`, i can't find one, so i accepted the change. i'll note that in the top comment.
> 
> other than that i did some other cleanups:
> 
>     1. added comment in the code explaining it's logic.
> 
>     2. remove unnecessary comment about internal implementation in the test
> 
> 
> since i already cloned the code and run the test locally, i did some other cleanup changes in the tests:
> 
>     * avoiding the use of "slave" when possible in new code
> 
>     * generalize client_field proc
> 
>     * move that test out of the scope of the server of the previous test.
> 
>     * speed up a previous test which i notices takes forever to finish.
> 
>     * negated the comment saying "and server cron to run and disconnect" (correct me if i'm wrong)
> 
> 
> let me know if you see anything wrong.

lgtm

sundb commented:
It seems that some CI failed due to this PR.
https://github.com/redis/redis/actions/runs/4400103760/jobs/7705087323

enjoy-binbin commented:
https://github.com/enjoy-binbin/redis/commit/2cbbf6394f015da0f2a40c8d521a87edb4858d18
This is what I suspect, I'm running Daily to verify it

sundb commented:
@enjoy-binbin Yes, we shouldn't have updated `master_repl` at the beginning, before this PR we would have created at most one new reply node, but now there are multiple.
The `repl_offset` indicates that the first byte of the reply node points to its position in the master reply, so when there are multiple nodes, it will always be wrong.

oranagra commented:
it doesn't really matters when we update `master_repl` since we also update `histlen`.
the problem is that now we're creating more than one node, and the `tail->repl_offset` of all the nodes except the last one are incorrect.
https://github.com/enjoy-binbin/redis/commit/2cbbf6394f015da0f2a40c8d521a87edb4858d18 looks right to me.

enjoy-binbin commented:
odd, my fix did not fix the CI failures. it still fails in the same problem
and look like we do not have the test that will cover my fix (tail->repl_offset one)

i will take another look later...

oranagra commented:
the problem, or at least one of them, is about NO_MALLOC_USABLE_SIZE, easy to reproduce locally with:
```
 make distclean; make MALLOC=libc CFLAGS=-DNO_MALLOC_USABLE_SIZE REDIS_CFLAGS='-Werror'
 ./runtest --single unit/maxmemory
```

xbasel commented:
> `repl_offset`

we probably need something liket this?


![image](https://user-images.githubusercontent.com/103044017/224649449-90d02263-9177-404d-a879-0ba647c9517b.png)

sundb commented:
Perhaps this is due to the overhead of the replBufBlock.

Following is the last log I printed, and I can see that jemalloc allocates 216 bytes more, so its `server.repl_buffer_blocks` length will be smaller, and `replBufBlock` overhead will be smaller.

jemalloc
```
malloc_size: 1024, usable_size: 1280, server.repl_buffer_blocks length: 40879
```

libc
```
malloc_size: 1024, usable_size: 1064, server.repl_buffer_blocks length: 49659
```

oranagra commented:
```c
            size_t size = min(max(len, (size_t)PROTO_REPLY_CHUNK_BYTES), (size_t)server.repl_backlog_size / 16);
```
in this test the backlog size is 10k, so backlog / 16 is smaller than PROTO_REPLY_CHUNK_BYTES. this seems to solve the problem, at least for this specific pair of tests.
```diff
-            size_t size = min(max(len, (size_t)PROTO_REPLY_CHUNK_BYTES), (size_t)server.repl_backlog_size / 16);
+                       size_t limit = max((size_t)server.repl_backlog_size / 16, PROTO_REPLY_CHUNK_BYTES);
+            size_t size = min(max(len, (size_t)PROTO_REPLY_CHUNK_BYTES), limit);
```

oranagra commented:
i suppose it's just a combination of a very low backlog size in that test, and some thresholds of that test that get violated because of the relatively high overhead of replBufBlock.
@enjoy-binbin can you please mix my fix with yours and try a full CI (maybe limit it to just the units we saw failing).

sundb commented:
@oranagra @enjoy-binbin Add tests to cover https://github.com/enjoy-binbin/redis/commit/2cbbf6394f015da0f2a40c8d521a87edb4858d18?

enjoy-binbin commented:
> can you please mix my fix with yours and try a full CI (maybe limit it to just the units we saw failing).

i am testing it, seems to work

> Add tests to cover xxx

yean, i'll give it a try, but maybe not in time
